### PR TITLE
V2: adapt CLI for multicomponent

### PR DESCRIPTION
**Pull request by hwpang**
## Description
This PR depends on https://github.com/chemprop/chemprop/pull/585. This PR adds support to train multicomponent models from CLI.

## Questions
- Is there a best time to add the `loc` and `scale` attributes to the predictor? Previously, I set them when the predictor object is constructed. However, I now realize that, in this case, the validation set should not be scaled. The alternative is to set them after the model is trained and compare them to the unscaled test set. I did the latter in this PR, but I am wondering if there is a preferred way.
- In this PR, I check if `n_components == 1` at multiple places to determine if it is monocomponenet or multicomponent. I welcome suggestions on a more elegant way to do this.

## Checklist
- [x] linted with flake8?

### CODE REVIEW

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -394,12 +395,12 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
...
-    #     "--split-key-molecule",
-    #     type=int,
-    #     default=0,
-    #     help="The index of the key molecule used for splitting when multiple molecules are present and constrained split_type is used, like scaffold_balanced or random_with_repeated_smiles.       Note that this index begins with zero for the first molecule.",
-    # )
+    split_args.add_argument(
+        "--split-key-molecule",
+        type=int,
+        default=0,
+        help="The index of the key molecule used for splitting when multiple molecules are present and constrained split_type is used, like scaffold_balanced or random_with_repeated_smiles.       Note that this index begins with zero for the first molecule.",
```
Comment (davidegraff) [inline chemprop/cli/train.py:402]:
```suggestion
        help="The index of the key molecule used for splitting when multiple molecules are present and constrained split_type is used (e.g., 'scaffold_balanced' or 'random_with_repeated_smiles'). Note that this index begins with zero for the first molecule.",
```

File: chemprop/cli/train.py
Comment (davidegraff) [inline chemprop/cli/train.py:1]: there are several checks in this file for multicomponent input and they are generally of the form `n_components == 1` or `n_components > 1`. When a condition is checked multiple times, it's best to give a name and reference that, because it makes logical checks more clear. For example:
```
if n_components > 1:
    # do something
else:
    # do something else
```
vs.
```
multicomponent = n_components > 1
if multicomponent:
    # do something
else:
    # do something else
```
This is legibility is compounded when you perform the same logical check many times because it tells a reader precisely what the `if`-statement is checking for rather than requiring a reader to parse it each time (i.e., when you want to check for monocomponent data, you just add a `not` rather than reversing the inequality).

File: chemprop/cli/predict.py [hunk 2]
```diff
@@ -302,14 +152,23 @@ def process_predict_args(args: Namespace) -> Namespace:
...
 
 def main(args):
-    model = MPNN.load_from_checkpoint(args.checkpoint_path)
 
-    bond_messages = isinstance(model.message_passing, BondMessagePassing)
+    if args.smiles_columns is None and args.reaction_columns is None:
+        n_components = 1
+    else:
+        n_components = len(args.smiles_columns) if args.smiles_columns is not None else 0
+        n_components += len(args.reaction_columns) if args.reaction_columns is not None else 0
```
Comment (davidegraff) [inline chemprop/cli/predict.py:165]:
```suggestion
    match (args.smiles_columns, args.reaction_columns):
        case [None, None]:
            n_components = 1
        case [_, None]:
            n_components = len(args.smiles_columns)
        case [None, _]:
            n_components = len(args.reaction_columns)
        case _:
            n_components = len(args.smiles_columns) + len(args.reaction_columns)
```
IMO this is cleaner to me. At the very least, split it up into `if`/`elif`/`else`

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -40,6 +40,13 @@ jobs:
         shell: bash -l {0}
         run: |
           pytest -v tests
+      - name: Test CLI
+        shell: bash -l {0}
+        run: |
+          chemprop train -i tests/data/regression/mol.csv --task-type regression --epochs 1 --num-workers 4
+          chemprop predict --checkpoint tests/data/example_model_v2.ckpt -i tests/data/regression/mol.csv --num-workers 4
+          chemprop train -i tests/data/regression/mol+mol.csv --task-type regression --smiles-columns smiles solvent --epochs 1 --num-workers 4
+          chemprop predict --checkpoint tests/data/example_model_v2_regression_multi.ckpt -i tests/data/regression/mol+mol.csv --smiles-columns smiles solvent --num-workers 4
```
Comment (davidegraff) [inline .github/workflows/tests.yml:49]: this is not how we should be testing the CI. It should be included in the integration tests under pytest and should be part of a separate PR

File: chemprop/cli/train.py [hunk 4]
```diff
@@ -547,8 +591,8 @@ def main(args):
 
     predictor_ffn = Factory.build(
         predictor_cls,
-        input_dim=mp_block.output_dim + train_dset.d_xf,
-        n_tasks=train_dset.Y.shape[1],
+        input_dim=mp_block.output_dim + train_dset.d_xf if n_components == 1 else mp_block.output_dim + sum(train_dset.datasets[i].d_xf for i in range(n_components)),
+        n_tasks=train_dset.Y.shape[1] if n_components == 1 else train_dset.datasets[0].Y.shape[1],
```
Comment (davidegraff) [inline chemprop/cli/train.py:595]:
```suggestion
        input_dim=mp_block.output_dim + d_xf),
        n_tasks=train_dset.Y.shape[1] if n_components == 1 else train_dset.datasets[0].Y.shape[1],
```
where `d_xf` is defined above:
```
d_df = train_dset.d_xf if n_components == 1 else mp_block.output_dim + sum(dset.d_xf for dset in train_dset.datasets)

File: chemprop/cli/train.py [hunk 5]
```diff
@@ -614,7 +661,8 @@ def main(args):
 
     if test_loader is not None:
         if args.task_type == "regression":
-            model.loc, model.scale = float(scaler.mean_), float(scaler.scale_)
+            model.predictor.register_buffer("loc", torch.tensor(scaler.mean_).view(-1, 1))
+            model.predictor.register_buffer("scale", torch.tensor(scaler.scale_).view(-1, 1))
```
Comment (davidegraff) [inline chemprop/cli/train.py:665]: I think right now we leave transparent output transform for a future feature. That is, we should _not_ do transformation inside the `Predictor` and instead opt to untransform using a separate `StandardScaler`.

File: pyproject.toml [hunk 6]
```diff
@@ -33,6 +33,7 @@ dependencies = [
     "rdkit",
     "scikit-learn",
     "scipy",
+    "tensorboard",
```
Comment (davidegraff) [inline pyproject.toml:36]: where in the code do we use `tensorboard`? It's a sub-dep of `lightning`, but this file only enumerates the libraries that are _directly_ needed.

File: chemprop/cli/predict.py (hunk 2 above)
Comment (hwpang) [reply chemprop/cli/predict.py:165]: I love this! Thanks for the suggestion!

File: pyproject.toml (hunk 6 above)
Comment (davidegraff) [reply pyproject.toml:36]: actually you're right. It looks like it's not installed, but is needed if we're using a `TensorboardLogger` in the CLI.

File: .github/workflows/tests.yml (hunk 3 above)
Comment (hwpang) [reply .github/workflows/tests.yml]: Thanks! I removed it here and opened an issue https://github.com/chemprop/chemprop/issues/590 to keep track of this.

File: chemprop/cli/train.py (hunk 5 above)
Comment (hwpang) [reply chemprop/cli/train.py:667]: @davidegraff Thanks for the comment! Wouldn't this requires users to have access to the training data, or maybe we need to add the `StandardScaler` in the checkpoint file somehow, so that users of the model can run inference and obtain what they expect?

Comment (davidegraff) [reply chemprop/cli/train.py:667]: not to the checkpoint (as it's not needed during training), but in the saved model output directory. I'm generally a fan of making these separate files:
```
output_no_cv/
├── model_0/
[...]
└── model_*
    ├── model.ckpt
    ├── preds.csv
    ├── input_scaler.pkl
    └── output_scaler.pkl
```

File: chemprop/cli/train.py
Comment (davidegraff) [reply chemprop/cli/train.py:1]: I think you might have mixed up many of your uses of the new `multicomponent` variable

**Comment by KnathanM** [comment]
Thank you for working on this Hao-Wei. I think we can discuss your two questions in our developer meeting on Monday and then get this merged in soon after.

File: chemprop/cli/train.py
Comment (hwpang) [reply chemprop/cli/train.py:1]: Oops, you're right, thanks for catching that. I will fix it.

File: chemprop/cli/train.py (hunk 5 above)
Comment (hwpang) [reply chemprop/cli/train.py:667]: I see, that makes sense. I will remove them from this PR and open an issue to keep track of it.

Comment (hwpang) [reply chemprop/cli/train.py:667]: Actually, I have a question. Regarding this part of code that test the model, if the transformation is not done inside the predictor, then we should transform the test set. The reported results here would be in the transformed space but not the original space with physical meaning. Is that what we intend to do?

```
    if test_loader is not None:
        if args.task_type == "regression":
            model.predictor.register_buffer("loc", torch.tensor(scaler.mean_).view(-1, 1))
            model.predictor.register_buffer("scale", torch.tensor(scaler.scale_).view(-1, 1))
        results = trainer.test(model, test_loader)[0]
        logger.info(f"Test results: {results}")
```

Comment (davidegraff) [reply chemprop/cli/train.py:667]: If I understand you correctly, the raw model output is in the standardized range, so you will need to use the `StandardScaler` to un-transform them back into the "physical" range. My general opinions which I think most people share:
- validation results should be comparable to training results, therefore we should **not** do any un-transformation
- testing results should be physically interpretable, therefore we _should_ un-transform.

**Review by davidegraff — APPROVED** [review]

File: chemprop/cli/predict.py [hunk 7]
```diff
@@ -186,7 +201,13 @@ def main(args):
         p_atom_descs=args.atom_descriptors_path,
         **featurization_kwargs,
     )
-    logger.info(f"test size: {len(test_data)}")
+
+    if n_components == 1:
```
Comment (KnathanM) [inline chemprop/cli/predict.py:205]: You could combine this logic check with the one on line 228 by waiting to pull test_data out of a list with something like:
```
logger.info(f"test size: {len(test_data[0])}")
...
test_dsets = [make_dataset(d, args.rxn_mode) for d in test_data]
if n_componenets == 1:
    test_dset = test_dsets[0]
else:
    test_dset = data.MulticomponentDataset(test_dsets)
```
not necessary, but a little simpler I think

Comment (hwpang) [reply chemprop/cli/predict.py]: Thanks! I like it!

File: chemprop/cli/train.py [hunk 8]
```diff
@@ -571,14 +614,17 @@ def main(args):
     train_loader = data.MolGraphDataLoader(train_dset, args.batch_size, args.num_workers)
     val_loader = data.MolGraphDataLoader(val_dset, args.batch_size, args.num_workers, shuffle=False)
     if len(test_data) > 0:
```
[the remark this answers is not in the document]
Comment (hwpang) [reply chemprop/cli/train.py:616]: Good catch! I will modify this

**Review by KnathanM — APPROVED** [review]
